src/callbacks/amr_dg1d.jl

```diff
@@ -0,0 +1,221 @@
+
+  # The "Ref(...)" is such that we can vectorize the search but not the array that is searched
+  elements_to_refine = searchsortedfirst.(Ref(cache.elements.cell_ids[1:nelements(dg, cache)]),
+                                          cells_to_refine)
+  needs_refinement[elements_to_refine] .= true
+
+  # Retain current solution data
+  old_n_elements = nelements(dg, cache)
+  old_u_ode = copy(u_ode)
+  GC.@preserve old_u_ode begin # OBS! If we don't GC.@preserve old_u_ode, it might be GC'ed
+    old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+    # Get new list of leaf cells
+    leaf_cell_ids = leaf_cells(mesh.tree)
+
+    # Initialize new elements container
+    elements = init_elements(leaf_cell_ids, mesh,
+                            real(dg), nvariables(equations), polydeg(dg))
+    copy!(cache.elements, elements)
+    @assert nelements(dg, cache) > old_n_elements
+
+    resize!(u_ode, nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache))
+    u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+    # Loop over all elements in old container and either copy them or refine them
+    element_id = 1
+    for old_element_id in 1:old_n_elements
+      if needs_refinement[old_element_id]
+        # Refine element and store solution directly in new data structure
+        refine_element!(u, element_id, old_u, old_element_id,
+                        adaptor, equations, dg)
+        element_id += 2^ndims(mesh)
+      else
+        # Copy old element data to new element container
+        @views u[:, .., element_id] .= old_u[:, .., old_element_id]
+        element_id += 1
+      end
+    end
+    @assert element_id == nelements(dg, cache) + 1 || element_id == nelements(dg, cache) + 2^ndims(mesh) "element_id = $element_id, nelements(dg, cache) = $(nelements(dg, cache))"
+  end # GC.@preserve old_u_ode
+
+  # TODO: Taal performance, allow initializing the stuff in place, making use of resize!
+  # Initialize new interfaces container
+  interfaces = init_interfaces(leaf_cell_ids, mesh, elements,
+                               real(dg), nvariables(equations), polydeg(dg))
+  copy!(cache.interfaces, interfaces)
+
+  # Initialize boundaries
+  boundaries, _ = init_boundaries(leaf_cell_ids, mesh, elements,
+                                  real(dg), nvariables(equations), polydeg(dg))
+  copy!(cache.boundaries, boundaries)
+
+  # Sanity check
+  if isperiodic(mesh.tree)
+    @assert ninterfaces(interfaces) == 1 * nelements(dg, cache) ("For 1D and periodic domains, the number of interfaces must be the same as the number of elements")
+  end
+
+  return nothing
+end
+
+
+# TODO: Taal compare performance of different implementations
+# Refine solution data u for an element, using L2 projection (interpolation)
+function refine_element!(u::AbstractArray{<:Any,3}, element_id, old_u, old_element_id,
+                         adaptor::LobattoLegendreAdaptorL2, equations, dg)
+  @unpack forward_upper, forward_lower = adaptor
+
```

Comment:
Here you seem to skip the `@boundscheck` that is part of the 2D implementation. Any reason for this?

Reply:
That's a relict of debugging AMR in 2D. I think we should have a discussion about our way of boundschecking in general. There are at least two possible ways:
1. Use more `@boundscheck` blocks and `@inbounds` afterwards to make the assumptions on bounds explicit and the code fast even if the users start Julia with default arguments. While debugging some code, we can start Julia with the flag `--check-bounds=yes` to override `@inbounds`.
2. Do not use `@boundscheck` or `@inbounds` an all. By default, Julia will check bounds on every access. Hence, users need to start Julia with `--check-bounds=no` (or `--check-bounds=0` in @sloede's custom build :wink:).

Reply:
IIRC, you tended towards the first variant. I would prefer the second one (for now), as it does not add another layer of complexity to our code. If at some point we do more large-scale production runs where debugging with bounds-checking enabled becomes too slow, we can revisit this question...

Reply:
Maybe we can use a hybrid way for now: If we want to introduce additional bounds checks, we hide them inside `@boundscheck` to eliminate them when using `--check-bounds=no`. What I have in mind is something like 
https://github.com/trixi-framework/Trixi.jl/blob/2c670e2b0f14e431428923ede8df90e130f9c158/src/amr/amr.jl#L22-L23
https://github.com/trixi-framework/Trixi.jl/blob/2c670e2b0f14e431428923ede8df90e130f9c158/src/solvers/dg/2d/amr.jl#L59-L62

Reply:
I've added the `@boundscheck` for 1D AMR in cdf6a78c75318db449786741ca1534083bf4ce63.

Reply:
Is cdf6a78c75318db449786741ca1534083bf4ce63 a valid hash (or have you pushed it yet)?

So, if I understand you correctly, you want to convert `@assert` to `@boundscheck` where some sort of "bounds" are concerned? In that case, I'd agree about converting the first one in amr/amr.jl, but the one in 2d/amr.jl is not so much a check on bounds but on whether there is no algorithmic error in the interface initialization routines. But if this qualifies as a boundscheck to you as well, go for it.

Reply:
`@boundscheck` tells Julia that the following code performs some bounds checks for security, see https://docs.julialang.org/en/v1/devdocs/boundscheck/. It doesn't perform a check like `@assert`. It's purpose is solely to mark some code as to be removed when bounds checking is disabled, e.g. by using `--check-bounds=no`. Hence, I don't want to convert `@assert stuff` to `@boundscheck stuff` but to `@boundscheck @assert stuff`.

I've just pushed the commit together with other ones to create only one new CI build.

Reply:
Maybe it's worth to create an issue for the `@boundscheck` stuff to discuss it further, also with the others.

Reply:
Tracked in #210.